api/src/main/java/org/apache/iceberg/Scan.java

```diff
@@ -129,6 +129,34 @@ default ThisT select(String... columns) {
    */
   ThisT planWith(ExecutorService executorService);
 
+  /**
+   * Create a new {@link TableScan} which dictates that when plan tasks via the {@link
+   * #planTasks()}, the scan should preserve partition boundary specified by the provided partition
+   * column names. In other words, the scan will not attempt to combine tasks whose input files have
+   * different partition data w.r.t `columns`.
+   *
+   * @param columns the partition column names to preserve boundary when planning tasks
+   * @return a table scan preserving partition boundary when planning tasks
+   * @throws IllegalArgumentException if any of the input columns is not a partition column, or if
+   *     the table is un-partitioned.
+   */
+  ThisT preservePartitions(Collection<String> columns);
+
+  /**
+   * Create a new {@link TableScan} which dictates that when plan tasks via the {@link
+   * #planTasks()}, the scan should preserve partition boundary specified by the provided partition
```

Comment:
If we were to guess, what attributes will Spark pass us during storage-partitioned joins in the future? Will it pass data attributes used in the join condition or actual partition columns? Seems like it would be join attributes. If so, should we accept a list of attributes instead of actual partition names?

Suppose I have a table partitioned by `day(ts), bucket(id)`. What's the input to this method?

Reply:
It'll pass the partition expressions to Iceberg, and is a subset of partition expressions Spark obtained via `SupportsReportPartitioning` from Iceberg, when the returned partitioning type is `KeyGroupedPartitioning`.

I think the partition expressions will be first processed via `SparkScanBuilder` (via a new interface to be introduced by Spark) and the actual partition columns will be passed to this method. Does that sound reasonable?

Reply:
Don't we need an already built `Scan` to obtain `KeyGroupedPartitioning`?

Reply:
Yea you're right. I think Spark can still push down partition expressions to Iceberg before a scan is built, and Iceberg uses that to create "adjusted" `KeyGroupedPartitioning` based on the actual partition columns used?

Reply:
@sunchao, what if we make Spark pass join keys instead of V2 partition expressions?

I think that can be easily done as the operator optimization batch in Spark runs before Spark builds scans, so we should know all equality predicates of our join condition by that time. That means, we can have a new mix-in interface for `ScanBuilder`. If our join condition is `t.dep = s.dep AND t.id = s.id` and tables are partitioned by `dep, bucket(id)`, Spark will pass just `dep`, `id` and Iceberg will interpret this as don't combine tasks across partition expressions on top `dep`, `id`.

Reply:
@aokolnychyi yes we can do that, in fact I think this is what we are currently proposing here: https://github.com/apache/spark/pull/38434. cc @huaxingao

Reply:
Cool, I think that PR is what we need.